PWM waveform, axi-pwmgen backport

threexc opened the pull request:
## PR Description

This PR includes several backports from upstream v6.12 and v6.13 branches around new PWM consumer APIs and waveform support, in preparation for the subsequent backport of the upstream ad7625 driver. Highlights:

- cherry-pick and cleanup rename of `pwm_apply_state()` to `pwm_apply_might_sleep()`
- backport v6.12 versions of `pwm/core.c`, `include/linux/pwm.h`, `include/trace/events/pwm.h`
- remove `drivers/pwm/sysfs.c` (now incorporated in `drivers/pwm/core.c`)
- cherry-pick `devm_clk_rate_exclusive_get()` addition
- backport v6.12 version of `drivers/pwm/pwm-axi-pwmgen.c`
- Cherry-pick the following from v6.13:
  - Improved pwm locking
  - PWM waveform abstractions, consumer APIs
  - PWM waveform tracing
  - axi-pwmgen driver implementation using PWM waveforms
  - `pwm/core.c` symbol reordering
  - kernel doc additions for new pwm_ops
  - chardev support for PWM devices

This has been tested locally using a Zedboard with a customized BOOT.BIN to route PWM signals to a PMOD header and the libpwm tooling found at https://git.kernel.org/pub/scm/linux/kernel/git/ukleinek/libpwm.git/, which allows setting PWM waveform properties via the newly-added character device interface.

## PR Type
- [ ] Bug fix (a change that fixes an issue)
- [X] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [X] I have conducted a self-review of my own code changes
- [X] I have tested the changes on the relevant hardware
- [ ] I have updated the documentation outside this repo accordingly (if there is the case)

nunojsa commented:
The backport broke the pulsar driver. These amount of backports in subsystems is a bit unsettling but I'll keep an open mind (also because a lot of the new stuff we have in the pipeline depend on the new PWM stuff). Oh well, hopefully this does not bring too much churn in my next merge.

threexc commented:
Updated the PR with:

- List of extra drivers modified in the ADI tree in commit 5c2cc67
- Cherry-pick of patch adding `axi_pwmgen_ddata_from_chip()`
- David's patches the axi-pwmgen driver's default alignment behaviors and register numbering
- Clarification in each cherry-picked patch stating the branch and commit ID it was taken from
- New patch adding the `phase` property back to `struct pwm_state` to make `ltc2387.c` and `ad_pulsar.c` compile, with a comment about the change

For the last point, I can confirm they both build OK, but I don't have the specific hardware to test them on.

dlech commented:
> * chardev support for PWM devices

This isn't finalized upstream yet, so probably best to not backport (unless we _really_ need it?).

> * to make `ltc2387.c`

IIRC, this one should be similar to ad7625 that you worked on? So should be easy to convert to use the new waveform stuff.


> * and `ad_pulsar.c`

Having worked on similar chips (ad7944, ad4000), I have strong doubts that this actually is using the phase feature. There is only one PWM used by the driver, so it doesn't seem like phase is actually doing anything there. Likely we can just delete that line.

dlech commented:
Found one more driver using phase: ad4630. That one uses 2 PWMs so we should convert it to the waveform APIs.

threexc commented:
FYI: Will be returning to this shortly, just getting some patches for the ad4695 driver ready to go upstream first.

threexc commented:
Rebased, not ready for review again yet.

threexc commented:
Rebased again, removed the old patch that preserved the phase field, and added three new patches to adjust the `ad_pulsar.c`, `ltc2387.c`, and `ad4630.c` drivers. Still need to test the latter, then I will make this ready for review.

threexc commented:
> > * chardev support for PWM devices
> 
> This isn't finalized upstream yet, so probably best to not backport (unless we _really_ need it?).

You're right, this probably isn't needed, but I was using it to test the basic PWM functionality. I'll remove it before marking this as ready for review again.

threexc commented:
This should be good to go now. I've tested with the AD4630 variants and a bare Zedboard with PWMs routed to the PMOD headers (although I have now removed the pwm chardev patch that allowed me to use libpwm to test this). 

ad_pulsar.c and ltc2387.c have been updated to use the PWM waveform API, but as mentioned before I'm not able to test these except to see that they compile.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -370,26 +372,28 @@ static int ad4630_read_avail(struct iio_dev *indio_dev,
-	conv_state.period = DIV_ROUND_CLOSEST(NSEC_PER_SEC, freq);
-	ret = pwm_apply_might_sleep(st->conv_trigger, &conv_state);
+	conv_wf.period_length_ns = DIV_ROUND_CLOSEST(NSEC_PER_SEC, freq);
+	ret = pwm_round_waveform_might_sleep(st->conv_trigger, &conv_wf);
```
If there is no retry on failure, then why call `pwm_round_waveform_might_sleep()` and not just `pwm_set_waveform_might_sleep()` directly? (If there is a good reason, please add a comment in the code to explain.)

threexc replied:
That's a good point. I don't think there's any value to doing that if we're not going to implement some retry/redirect based on the rounding. I saw this combo mentioned in the comment for `pwm_round_waveform_might_sleep()` and just started implementing it everywhere. The driver is fine without it from my testing, so I'll just use `pwm_set_waveform_might_sleep()`.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -398,23 +402,33 @@ static int __ad4630_set_sampling_freq(const struct ad4630_state *st, unsigned in
+	 * so round this setting to the closest available value.
 	 */
-	fetch_state.phase = AD4630_TQUIET_CNV_DELAY_PS;
+	fetch_wf.duty_offset_ns = DIV_ROUND_CLOSEST_ULL(AD4630_TQUIET_CNV_DELAY_PS, KILO);
```
How about just changing the macro to `AD4630_TQUIET_CNV_DELAY_NS` so we don't have to divide here?

threexc replied:
It'll have to be rounded up to an even 10, but yes that works too.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -398,23 +402,33 @@ static int __ad4630_set_sampling_freq(const struct ad4630_state *st, unsigned in
+	if (ret)
+		return ret;
+
+	st->conv_wf = conv_wf;
+	st->fetch_wf = fetch_wf;
```
Shouldn't these be after the `pwm_set_...` in case if fails?

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -756,6 +790,8 @@ static int ad4630_buffer_postdisable(struct iio_dev *indio_dev)
 out_error:
 	pm_runtime_mark_last_busy(&st->spi->dev);
 	pm_runtime_put_autosuspend(&st->spi->dev);
+	pwm_disable(st->fetch_trigger);
+	pwm_disable(st->conv_trigger);
```
Not sure about this since it wasn't there before.

If ever used with a different PWM controller, the output pin could do anything when disabled, which could be problematic.

threexc replied:
I suppose so. I was following what I did for the ad7625 upstream, but it's another case of it works fine without it and removing this makes it that much closer to what it was, so I'll make that change.

dlech commented:
We could probably reduce future merge conflicts by adding `#define pwm_apply_state pwm_apply_might_sleep` to `pwm.h` instead of backporting the rename commit that didn't apply cleanly.

threexc commented:
Any updates? I see that there are some errors in the CI builds, but they all seem to be about things that are actually present upstream. Do we want to fix those?

dbogdan commented:
Let's wait until the update from Linux Kernel v6.1 to v6.6 on the main branch will be completed (it should happen in the next few days). @nunojsa is actively working on this.

nunojsa commented:
Yeah, seems there are some conflicts... But I see now these backports would have been useful for the 6.6 rpi branch and the automatic cherry-picking will likely fail. Oh well, it is what it is... I'll fix them manually.

threexc commented:
This is on my radar to fix today, so I'll see if I can resolve the conflicts.

threexc commented:
It was actually pretty easy to resolve after the first few patches. I still need to test it out on my Zedboard - setting that up now.

threexc commented:
Seems to be working OK with the latest rebase on 6.6. I tested it locally on a Zedboard with the AD7625 EVAL on top (those specific driver patches are on a different branch, but if they work it means the underlying PWM logic is functioning).

Should be good for final review and (maybe) a merge.

nunojsa commented on drivers/iio/adc/ad4630.c:
```diff
@@ -731,6 +742,14 @@ static int ad4630_buffer_preenable(struct iio_dev *indio_dev)
+
+	ret = pwm_set_waveform_might_sleep(st->fetch_trigger, &st->fetch_wf, false);
+	if (ret)
+		return ret;
```
This looks nok to me... Above we have ad4630_sampling_enable() which seems not to be doing a thing. AFAIU from the new API, `pwm_set_waveform_might_sleep()` will enable the PWM output if we have a non zero period, right? Even the above calls look unnecessary to me... So, if I'm not missing nothing, I think this poses an opportunity to change things a bit... 

As we had before, it also did not made too much sense as we only need to enable the PWM when buffering is ON. So either:

1. We only cache the frequency and apply the waveform on buffer preenable. And we should error out in case frequency is set to 0.
2. We apply the waverform without enabling the output. But this one would need a new API in the pwm subsystem I guess? Or maybe do pwm_disable() after applying the waveform...

What is not that nice with 1) is that if we fail to apply the waveform we'll only know about it when actually enabling the IIO buffer which can be a bit odd (one would expect that if writing the sampling frequency worked, then we should be fine to go). 2) needs a new API but I think lots of the IIO users are like this... they want to apply a waveform but only enable it when buffering starts.

A third option is just to continue as of today and enable the waveform right away when we set the frequency. But then we do not need any of the above calls (including ad4630_sampling_enable()). Instead we would need to get the pwm state and error out if disabled.

Does it make sense? I might be missing something on how the new API is supposed to work...

dlech replied:
Technically, `pwm_disable()` could change the output to anything for an arbitrary PWM controller. So what I have been trying to do so far is to set the PWM period (so that we don't have to cache it) and use the duty cycle to turn the the output on or off. I.e. the set sample rate function only changes the period and not the duty cycle and the buffer enable/disable functions only set the duty cycle, not the period.

threexc commented:
Pushed a rebase. Haven't included ad4630 fixes yet.

threexc commented:
@nunojsa @dlech let me know what you think of the latest version - this makes it behave similarly to how the ad7625 does upstream (and eventually in the backport), but I've omitted calls to `pwm_round_waveform_might_sleep()` since the `set()` calls will do the rounding too.

Still trying to avoid making too many changes compared to baseline, but if you think it needs to be more drastic I can do that.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -258,10 +261,10 @@ static int ad4630_reg_access(struct iio_dev *indio_dev, unsigned int reg,
 
-	pwm_get_state(st->conv_trigger, &conversion_state);
-	*freq = DIV_ROUND_CLOSEST_ULL(NANO, conversion_state.period);
+	pwm_get_waveform_might_sleep(st->conv_trigger, &conversion_wf);
```
This is not valid when the PWM is disabled. Instead, we can just use the cached `st->conv_wf`.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -370,26 +373,21 @@ static int ad4630_read_avail(struct iio_dev *indio_dev,
-	ret = pwm_apply_state(st->conv_trigger, &conv_state);
-	if (ret)
-		return ret;
+	conv_wf.period_length_ns = DIV_ROUND_CLOSEST(NSEC_PER_SEC, freq);
```
We need to replace the calls to `pwm_apply_state()` with `pwm_round_waveform_might_sleep()` to validate that the PWM can actually do what is requested. With the added bonus that the period in `conv_wf` will be the actual period that can be produced and not just the requested period.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -592,17 +596,17 @@ static int ad4630_write_raw(struct iio_dev *indio_dev,
-	pwm_get_state(st->fetch_trigger, &fetch_state);
-	fetch_state.period = conv_state.period * 1 << avg;
-	fetch_state.phase = AD4630_TQUIET_CNV_DELAY_PS;
+	pwm_get_waveform_might_sleep(st->conv_trigger, &conv_wf);
```
Shouldn't this be using the saved copies of `conv_wf` and `fetch_wf` from `__ad4630_set_sampling_freq()`?

The old `pwm_get_state()` gets the _requested_ state, but `pwm_get_waveform_might_sleep()` gets the _actual_ hardware state. If this is called when PWM is disabled, then the period will be 0 and this won't work as expected.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -592,17 +596,17 @@ static int ad4630_write_raw(struct iio_dev *indio_dev,
+	fetch_wf.duty_offset_ns = AD4630_TQUIET_CNV_DELAY_NS;
 
-	return pwm_apply_state(st->fetch_trigger, &fetch_state);
+	return pwm_set_waveform_might_sleep(st->fetch_trigger, &fetch_wf, false);
```
Do we actually want to enable the PWM here or just call `pwm_round_waveform_might_sleep()`? Also, don't we want to update `st->fetch_wf` on success?

threexc replied:
Yeah you're right about all of these things. Fixup incoming.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -370,26 +370,24 @@ static int ad4630_read_avail(struct iio_dev *indio_dev,
-		.duty_cycle = 10,
-		.enabled = true,
+	struct pwm_waveform conv_wf = {
+		.duty_length_ns = 10,
```
Would be safer to check rounded waveform to ensure this minimum time is met since it could be rounded down. Or pick a higher value that can be safely rounded down, if that is even possible.

threexc replied:
I've added a warning in the changes I'm about to push - maybe it should be a full error, instead?

dlech replied:
Neither. One of the reasons for the new waveform API is so that we can keep retrying until we get a good value. There should not be an error unless it is impossible for the hardware to get a good value.

Another alternative would be that instead of requesting the smallest possible duty cycle, we request 50%, assuming that is within the datasheet specs. Or we request the largest allowable duty cycle instead of the smallest since it is rounding down.

threexc replied:
The datasheet doesn't specify a maximum, so what I've done is added a while loop in the function to check if `duty_length_ns` is less than 10, incrementing the target value by 10 and retrying if it's not. I've added similar logic for the fetch waveform's `duty_offset_ns` since that one is currently just taking the rounded result without checking, too.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -590,26 +598,32 @@ static int ad4630_write_raw(struct iio_dev *indio_dev,
+static int ad4630_update_sample_fetch_trigger(struct ad4630_state *st, u32 avg)
 {
-	struct pwm_state fetch_state, conv_state;
+	struct pwm_waveform fetch_wf = { };
```
```suggestion
	struct pwm_waveform fetch_wf = st->fetch_wf;
```

Then we don't need to add `fetch_wf.duty_length_ns = st->fetch_wf.duty_length_ns;`.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -729,7 +724,17 @@ static int ad4630_buffer_preenable(struct iio_dev *indio_dev)
+
+	ret = pwm_set_waveform_might_sleep(st->conv_trigger, &st->conv_wf, false);
+	if (ret)
+		return ret;
```
Don't these need to `goto out_error`? And also disable `conv_trigger` if enabling `fetch_trigger` fails. (and don't forget to add `spi_unoptimize_message()` to the error unwind stack.

Also, existing bug that we don't re-enter config mode on error after `regmap_write(st->regmap, AD4630_REG_EXIT_CFG_MODE, BIT(0));`

ukleinek commented on drivers/iio/adc/ltc2387.c:
```diff
@@ -186,12 +185,12 @@ static int ltc2387_set_sampling_freq(struct ltc2387_dev *ltc, int freq)
+	cnv_wf.period_length_ns = div_u64_rem((u64)DIV_ROUND_UP(ltc->ref_clk_rate, freq) * NSEC_PER_SEC,
+					      ltc->ref_clk_rate, &rem);
 	if (rem)
-		cnv_state.period += 1;
+		cnv_wf.period_length_ns += 1;
 
-	ret = pwm_apply_state(ltc->cnv, &cnv_state);
+	ret = pwm_round_waveform_might_sleep(ltc->cnv, &cnv_wf);
```
Ideally this would use the round function provided by the pwm framework. This needs some helper functions first though before it can become usable, so I just made a mental note for now.

ukleinek commented on drivers/iio/adc/ltc2387.c:
```diff
@@ -201,14 +200,11 @@ static int ltc2387_set_sampling_freq(struct ltc2387_dev *ltc, int freq)
+	clk_gate_wf.duty_offset_ns = LTC2387_T_FIRSTCLK_NS;
 
-	ret = pwm_apply_state(ltc->clk_en, &clk_en_state);
+	ret = pwm_round_waveform_might_sleep(ltc->clk_en, &clk_gate_wf);
```
`s/pwm_round_waveform_might_sleep/pwm_set_waveform_might_sleep/`, or am I missing something here?

threexc replied:
Thanks for catching that. This driver is very similar to the upstream ad7625 one (it's actually the template I used for it). There, we only round the waveforms in the `set_sampling_freq()` function and reserve enabling/disabling the PWM for the `buffer_preenable()` and `buffer_postdisable()` functions, but those don't exist here, so using the set function here probably makes more sense...

ukleinek commented on drivers/iio/adc/ad4630.c:
```diff
@@ -398,23 +413,42 @@ static int __ad4630_set_sampling_freq(const struct ad4630_state *st, unsigned in
+		fetch_wf.duty_offset_ns = target;
+		ret = pwm_round_waveform_might_sleep(st->fetch_trigger, &fetch_wf);
+		if (ret)
+			return ret;
```
Note that this loop probably changes `fetch_wf.duty_offset_ns`.

ukleinek replied:
I misread the code. Looks fine enough. (To make it completely right is too complicated here. The respective code should be in a pwm helper.) I think you could simplify the code a bit, to something like:
```
do {
   fetch_wf.duty_offset_ns = target;
   ret = pwm_round_waveform_might_sleep(..., &fetch_wf);
   if (ret)
      return ret;
   target += 10;
} while (fetch_wf.duty_offset_ns < AD4630_TQUIET_CNV_DELAY_NS)
```

ukleinek commented on drivers/iio/adc/ad4630.c:
```diff
@@ -590,26 +624,31 @@ static int ad4630_write_raw(struct iio_dev *indio_dev,
-	pwm_get_state(st->fetch_trigger, &fetch_state);
-	fetch_state.period = conv_state.period * 1 << avg;
-	fetch_state.phase = AD4630_TQUIET_CNV_DELAY_PS;
+	fetch_wf.period_length_ns = st->conv_wf.period_length_ns * 1 << avg;
```
`A * 1 << B == A << B`

threexc commented on include/linux/pwm.h:
```diff
@@ -6,6 +6,10 @@
 #include <linux/mutex.h>
 #include <linux/of.h>
 
+MODULE_IMPORT_NS(PWM);
```
Ah, I missed that while working through a rebase. It's in the upstream file, so I'll move it into the subsequent patch.

threexc commented on drivers/clk/clk.c:
```diff
@@ -940,6 +940,25 @@ int clk_rate_exclusive_get(struct clk *clk)
+	clk_rate_exclusive_put(clk);
+}
+
+int devm_clk_rate_exclusive_get(struct device *dev, struct clk *clk)
+{
+	int ret;
+
+	ret = clk_rate_exclusive_get(clk);
+	if (ret)
+		return ret;
+
+	return devm_add_action_or_reset(dev, devm_clk_rate_exclusive_put, clk);
+}
+EXPORT_SYMBOL_GPL(devm_clk_rate_exclusive_get);
```
There's no `devm_clk_rate_exclusive_get()` in the pwmgen driver? At least not in this series.

dlech replied:
It is in this series, being removed as is should be. :wink: 

https://github.com/analogdevicesinc/linux/pull/2640/files#diff-f0b252602282be62ee65e7a75dd0aa0c1cf2ce7f901acd0d074fb7e1d960affbL84

threexc replied:
I was wondering why I couldn't find it in a search of the current code!

threexc commented:
Addressed everything except for the `devm_clk_rate_exclusive_get()` comment - don't see the issue mentioned there.

threexc commented:
Looks like there are more issues with other PWM devices I'll have to fix.

threexc commented:
@ukleinek was kind enough to help put together a more extensive PWM backport from upstream in December, which I've now pushed and thrown my last few phase-related customizations on top of. I've tested this with an ad4630-24 again and it works, and I haven't seen any build issues either.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -370,26 +370,41 @@ static int ad4630_read_avail(struct iio_dev *indio_dev,
+	 * rounding is severe enough to keep putting the first few results under
+	 * the minimum 10ns condition checked by the while loop.
+	 */
+	ret = pwm_round_waveform_might_sleep(st->conv_trigger, &conv_wf);
```
This no longer has an effect because the while loop below unconditionally calls `pwm_round_waveform_might_sleep()` again. So we can remove it.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -370,26 +370,41 @@ static int ad4630_read_avail(struct iio_dev *indio_dev,
+		.duty_length_ns = 10,
 	};
 	int ret;
+	u64 target = conv_wf.duty_length_ns;
```
This is a roundabout way to set `target` to 10. Later, we unconditionally set `conv_wf.duty_length_ns = target`, so there isn't much point initializing `conv_wf` with the same value before that.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -398,23 +413,42 @@ static int __ad4630_set_sampling_freq(const struct ad4630_state *st, unsigned in
+	 * so round this setting to the closest available value.
 	 */
-	fetch_state.phase = AD4630_TQUIET_CNV_DELAY_NS;
+	fetch_wf.duty_offset_ns = AD4630_TQUIET_CNV_DELAY_NS;
+	target = fetch_wf.duty_offset_ns;
+
+	ret = pwm_round_waveform_might_sleep(st->fetch_trigger, &fetch_wf);
+	if (ret)
+		return ret;
+
+	do {
+		fetch_wf.duty_offset_ns = target;
```
Similar comments here. `fetch_wf.duty_offset_ns` is unnecessarily initialized to 10 before this. And it is extra work to set `fetch_wf.duty_offset_ns = AD4630_TQUIET_CNV_DELAY_NS;`, then `target = fetch_wf.duty_offset_ns;` then `fetch_wf.duty_offset_ns = target;` and we also have the extra call to `pwm_round_waveform_might_sleep()` before the while loop.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -590,26 +624,31 @@ static int ad4630_write_raw(struct iio_dev *indio_dev,
-	fetch_state.period = conv_state.period * 1 << avg;
-	fetch_state.phase = AD4630_TQUIET_CNV_DELAY_NS;
+	fetch_wf.period_length_ns = st->conv_wf.period_length_ns << avg;
+	fetch_wf.duty_offset_ns = AD4630_TQUIET_CNV_DELAY_NS;
```
I don't think we want to touch duty cycle here (I'm making the assumption it is always already set correctly in `__ad4630_set_sampling_freq()` before this - but may not be a valid assumption, I didn't look too hard.)

Otherwise we will the the same while loop to ensure a valid duty cycle value.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -708,6 +728,7 @@ static int ad4630_buffer_preenable(struct iio_dev *indio_dev)
 {
 	struct ad4630_state *st = iio_priv(indio_dev);
 	int ret;
+	u32 dummy;
```
`s/dummy/unused/` would make the intention a bit more clear IMHO.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -719,8 +740,11 @@ static int ad4630_buffer_preenable(struct iio_dev *indio_dev)
-	if (ret)
+	if (ret) {
+		/* make sure we can still access the registers */
+		regmap_read(st->regmap, AD4630_REG_ACCESS, &dummy);
```
We aren't checking the return value or `dummy`, so how is this making sure we can still access registers?

Seems like an unrelated change to the PWM stuff anyway, so deserves a separate commit.

threexc replied:
This is mirroring other parts of the driver where a read is done on that register to enter register configuration mode:

```
After writing to or reading from the appropriate user registers,
the host controller must exit the register configuration mode by
writing 0x01 to Register Address 0x0014 as detailed in the exit
configuration mode register. An algorithm for register read/write
access is as follows:
1. Perform a readback from dummy Register Address 0x3FFF to
enter the register configuration mode.
2. Readback from or write to the desired user register addresses.
3. Exit the register configuration mode by writing 0x01 to Register
Address 0x0014. Exiting register configuration mode causes the
register updates to take effect.
```

The issue here is we don't want to be setting ret (because if regmap_read() returns 0 then the report from out_error will be confusing, since it's supposed to be an error). 

I guess I could use a second variable to return from inside the if block?

dlech replied:
>  read is done on that register to enter register configuration mode
 
Got it. If the comment said something along the lines of "reading this register returns to config mode" would make it more clear.

> I guess I could use a second variable to return from inside the if block?

Yeah, not much we could do with it other than log it though since it is on the error return path.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -729,12 +753,25 @@ static int ad4630_buffer_preenable(struct iio_dev *indio_dev)
 out_error:
 	pm_runtime_mark_last_busy(&st->spi->dev);
 	pm_runtime_put_autosuspend(&st->spi->dev);
+	spi_unoptimize_message(&st->offload_msg);
```
This should be before `out_error:` and have a new label, otherwise errors before `spi_optimize_message()` will incorrectly call `spi_unoptimize_message()`.

dlech reviewed (commented):
I assume everything other than the last 3 commits is cherry-picked from upstream?

threexc commented:
Pushed review-based updates to the ad4630 driver. No other changes.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -370,26 +370,33 @@ static int ad4630_read_avail(struct iio_dev *indio_dev,
-		.duty_cycle = 10,
-		.enabled = true,
-	};
+	struct pwm_waveform conv_wf = { }, fetch_wf = { .duty_length_ns = 10 };
```
```suggestion
	struct pwm_waveform conv_wf = { }, fetch_wf = { };
```

This gets unconditionally written over later.

threexc replied:
It's actually duty_offset that gets written over. If I don't set this here, the buffer fails to initialize during testing.

dlech replied:
My bad.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -370,26 +370,33 @@ static int ad4630_read_avail(struct iio_dev *indio_dev,
-	};
+	struct pwm_waveform conv_wf = { }, fetch_wf = { .duty_length_ns = 10 };
 	int ret;
+	u64 target = 10;
```
Can we use `AD4630_TQUIET_CNV_DELAY_NS` here? Or introduce a new macro?

threexc replied:
Added a new AD4630_TCNV_HIGH_NS macro for it.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -398,23 +405,37 @@ static int __ad4630_set_sampling_freq(const struct ad4630_state *st, unsigned in
+	 * so round this setting to the closest available value.
 	 */
-	fetch_state.phase = AD4630_TQUIET_CNV_DELAY_NS;
+	target = AD4630_TQUIET_CNV_DELAY_NS;
```
Previously, `AD4630_TQUIET_CNV_DELAY_NS` was for phase, not duty. They have the same value, but is it actually the same thing when looking at timing diagrams in the datasheet?

Might need a new macro, or just stick with the hard-coded 10 as it was before.

dlech replied:
Based on your other reply, I'm guessing this comment should be ignored as well because I missed the duty_cycle vs. duty_offset.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -719,20 +726,39 @@ static int ad4630_buffer_preenable(struct iio_dev *indio_dev)
 
 	ret = regmap_write(st->regmap, AD4630_REG_EXIT_CFG_MODE, BIT(0));
-	if (ret)
+	if (ret) {
 		goto out_error;
+	}
```
Unnecessary change.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -719,20 +726,39 @@ static int ad4630_buffer_preenable(struct iio_dev *indio_dev)
 	ret = spi_optimize_message(st->spi, &st->offload_msg);
 	if (ret < 0)
-		goto out_error;
+		goto out_unoptimize;
```
Don't change this. If we failed to optimize, then we should not unoptimize.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -719,20 +726,39 @@ static int ad4630_buffer_preenable(struct iio_dev *indio_dev)
+
+	ret = pwm_set_waveform_might_sleep(st->fetch_trigger, &st->fetch_wf, false);
+	if (ret) {
+		pwm_disable(st->conv_trigger);
```
Might as well add `pwm_disable()` to the error unwind path with a new label to keep the pattern consistent.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -741,19 +767,18 @@ static int ad4630_buffer_preenable(struct iio_dev *indio_dev)
 {
 	struct ad4630_state *st = iio_priv(indio_dev);
-	u32 dummy;
+	u32 unused;
```
Oh, didn't realize this naming was already used thought the driver. Maybe better to just leave it then.

nunojsa commented:
> @ukleinek was kind enough to help put together a more extensive PWM backport from upstream in December, which I've now pushed and thrown my last few phase-related customizations on top of. This solves a bunch of build issues we saw with my previous version, which worked for the PWM drivers but broke a lot of other things. I've tested this with an ad4630-24 again and it works, and I haven't seen any build issues either.

I have to be honest... this is a bit concerning. We have now 250 patches in the backport. Was there any major/blocker issue with the previous approach that justifies this? I'm foreseeing lot's of pain and conflicts next time we merge to another kernel (even more if the next version we go happens to not include the waveform API).

threexc commented:
> > @ukleinek was kind enough to help put together a more extensive PWM backport from upstream in December, which I've now pushed and thrown my last few phase-related customizations on top of. This solves a bunch of build issues we saw with my previous version, which worked for the PWM drivers but broke a lot of other things. I've tested this with an ad4630-24 again and it works, and I haven't seen any build issues either.
> 
> I have to be honest... this is a bit concerning. We have now 250 patches in the backport. Was there any major/blocker issue with the previous approach that justifies this? I'm foreseeing lot's of pain and conflicts next time we merge to another kernel (even more if the next version we go happens to not include the waveform API).

The main motivating issue was that with a minimal port of only the axi-pwmgen driver and the core all other drivers are broken. So if a user of the ADI vendor tree wants to use a PWM for a non-ADI related purpose, they have to fix this driver first. This is an experience we don't want a user to suffer from.

For the record, the failures seen when trying to run `make drivers/pwm/` with the original backport tend to look like:

```
drivers/pwm/pwm-crc.c: In function 'crc_pwm_apply':
drivers/pwm/pwm-crc.c:58:30: error: incompatible types when initializing type 'struct device *' using type 'struct device'
   58 |         struct device *dev = crc_pwm->chip.dev;
      |                              ^~~~~~~
drivers/pwm/pwm-crc.c: In function 'crc_pwm_get_state':
drivers/pwm/pwm-crc.c:128:30: error: incompatible types when initializing type 'struct device *' using type 'struct device'
  128 |         struct device *dev = crc_pwm->chip.dev;
      |                              ^~~~~~~
drivers/pwm/pwm-crc.c: In function 'crystalcove_pwm_probe':
drivers/pwm/pwm-crc.c:171:25: error: incompatible types when assigning to type 'struct device' from type 'struct device *'
  171 |         pwm->chip.dev = &pdev->dev;
      |                         ^

```
Additionally there are some conflicts that are not detected by git and only break at compile time or even at runtime only.

Further we (=@threexc and @ukleinek) expect that an update to a newer kernel will be smooth as the commits pretty much match mainline and so most of them just drop out cleanly (or are kept cleanly if they are newer in mainline than the new base version) and so yields less friction than a custom partial port even.
If the next update will be to something older than 6.13, there is a trick to reduce conflicts for merges (a rebase should be fine as described above). If you're interested we can elaborate. And we gladly support you when the time has come and you need a helping hand with conflict resolution.

So even if the patch count is big, we're convinced this is the right balance between effort and benefit for ADI and value for ADI's customers.

nunojsa commented:
> The main motivating issue was that with a minimal port of only the axi-pwmgen driver and the core all other drivers are broken. So if a user of the ADI vendor tree wants to use a PWM for a non-ADI related purpose, they have to fix this driver first. This is an experience we don't want a user to suffer from.

Agreed... The other drivers should still build even if we do not use them and we had a recent case (on pwm-cadence) in the 2023_R2 release.

> If the next update will be to something older than 6.13, there is a trick to reduce conflicts for merges (a rebase should be fine as described above). If you're interested we can elaborate

That's what I fear... Yeah, we can't really rebase on main but I do normally have a rebased branch which needs to have a NULL diff with main and can be used to make sure `git merge` did not silently broke any driver. On the other hand, there are good chances we'll go to something higher or equal as 6.13 as there's still no longterm version after 6.6 (and xilinx typically uses longterm).

Alright, let's see how this works out. Will wait from an ack from David ou Uwe since they were the most active (together with you) in these process.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -729,10 +737,27 @@ static int ad4630_buffer_preenable(struct iio_dev *indio_dev)
+		goto out_unoptimize;
 
 	return 0;
+out_unoptimize:
```
Looks like the error unwinding still needs some work. Things should be in the exact reverse order of which they were set up. And more labels are needed to make sure we aren't undoing things that weren't done.

threexc replied:
I've added another label and did some reordering. I think it should be correct now.

dlech reviewed (commented):
"iio: adc: ltc2387.c: update to use PWM waveforms" could use a commit description. 

I only looked at the last 3 patches that aren't backports since I assume that Uwe has already looked at the others that are being backported.

threexc commented on drivers/iio/adc/ad4630.c:
```diff
@@ -720,19 +728,37 @@ static int ad4630_buffer_preenable(struct iio_dev *indio_dev)
+
+	ret = pwm_set_waveform_might_sleep(st->conv_trigger, &st->conv_wf, false);
+	if (ret)
+		goto out_unoptimize;
```
It seems like this means we don't need an `out_unoptimize` label then, since nothing goes directly to it. Unless I missed something?

dlech replied:
Not unless you fix the missing error check on `spi_engine_ex_offload_load_msg()`. :wink:

But we can save that for a follow up patch.

dlech replied:
Correction, still should not have `out_unoptimize`, otherwise we fail to release the mutex lock.

threexc replied:
Just dropping that patch.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -744,16 +773,17 @@ static int ad4630_buffer_postdisable(struct iio_dev *indio_dev)
 	spi_unoptimize_message(&st->offload_msg);
 	ret = regmap_read(st->regmap, AD4630_REG_ACCESS, &dummy);
-out_error:
+	if (ret)
+		return ret;
```
This is like the error unwinding, so we don't really want to return here, otherwise pm runtime stuff will not have balanced calls. 

Things are going to be in a bad state either way though.

threexc replied:
Changed it to a warning like the other case. Also added a patch to fix the load_msg() call not being handled.

nunojsa commented on drivers/iio/adc/ad4630.c:
```diff
@@ -744,16 +773,17 @@ static int ad4630_buffer_postdisable(struct iio_dev *indio_dev)
 	ret = regmap_read(st->regmap, AD4630_REG_ACCESS, &dummy);
-out_error:
+	if (ret)
+		pr_warn("%s:%d: couldn't reenter register configuration mode\n", __func__, __LINE__);
```
why pr_warn() and not dev_warn()?

threexc replied:
Good point, I use pr_warn() for debug but dev_warn() is preferred when a `dev` object is available. Pushed a change for that.

nunojsa reviewed (commented):
Ok, it seems this is mostly ready t go. Just one minor thing I'd like to be fixed...
